documentation/projects/proposals/search_relevancy_sandbox/20230530-implementation_plan_staging_elasticsearch_reindex_dags.md

```diff
@@ -0,0 +1,192 @@
+- [Project Proposal](./20230331-project_proposal_search_relevancy_sandbox.md)
+
+## Overview
+
+This document describes the addition of two DAGs for Elasticsearch (ES) index
+creation ––full and proportional-by-provider–– which will allow us to decouple
+the process from the long Ingestion server's data refresh process and experiment
+with smaller indices. Also includes the adoption of two new index aliases for
+ease of handling and querying the new index types from the API with the
+[`internal__index`](api_ii_param) param.
+
+[api_ii_param]: https://github.com/WordPress/openverse/pull/2073
+
+## Expected Outcomes
+
+- 1 DAG for fully recreation of the index that the API uses which includes the
+  includes all the database contents.
+- 1 DAG to create indexes of reduced size but
+  proportional-to-production-by-provider index.
+
+## Dependencies
+
+Same as for
+[Implementation Plan: Update Staging Database](./20230406-implementation_plan_update_staging_database.md).
+
+This work is related to the
+[Staging database recreation DAG](staging_db_recreation) plan for having
+production volumes in the staging DB, which is expected to finish soon.
+
+[staging_db_recreation]: https://github.com/WordPress/openverse/issues/1989
+
+## DAGs
+
+### `recreate_full_<media>_index` DAG
+
+#### Parameters
+
+1. `media_type`: The media type for which the index is being created. Presently
+   this would only be `image` or `audio`.
+2. `promote`: (Optional) A boolean value to indicate if result index will
+   replace the current one pointed by the `media` alias. If `True`, then the new
+   index will be the one used by the staging API. Defaults to `False`.
+3. `delete_old_if_promoted`: (Optional) A boolean value to indicate if the old
+   index pointed by the `<media>` alias should be deleted after replacement.
+   Defaults to `False`.
+
+#### Outlined Steps
+
+This DAG will leverage the **Ingestion server's API** and use the existing
+[`REINDEX` task](REINDEX), which is the same used by the data refresh process to
+create the index.
+
+1. Get the current timestamp in order to create the index suffix with the form
+   of `full-<timestamp>`.
+2. Use the `ingestion_server.trigger_and_wait_for_task()` utility to send the
+   `REINDEX` call to ingestion server, passing the previously generated
+   `index_suffix` in the data payload.
+3. If `promoted=True` is passed, then inmediatly make the `media` alias point to
+   the new index. A
+4. If the index is promoted then the DAG checks if `remove_old_if_promoted=True`
+   and proceeds to trigger the [DELETE_INDEX](delete_index) task in the
+   Ingestion server. Otherwise the DAG ends at the previous step.
+
+[reindex]:
+  https://github.com/WordPress/openverse/blob/7427bbd4a8178d05a27e6fef07d70905ec7ef16b/ingestion_server/ingestion_server/indexer.py#L282
+[delete_index]:
+  https://github.com/WordPress/openverse/blob/7427bbd4a8178d05a27e6fef07d70905ec7ef16b/catalog/dags/data_refresh/data_refresh_task_factory.py#L222-L239
```

Comment:
(note that it looks like this comment was cut off Sara)

That's an interesting suggestion! I think that the additional complexity around preventing deletion & managing concurrency might not make this approach worth it, even if it means relying on the ingestion server less.

Reply:
For what it's worth we already have complex concurrency rules for the same reasons in DAGs that use the ingestion server, so it wouldn't necessarily increase overall complexity of managing these things. Staging ingestion server only has two workers, so it'll likely go a _lot_ slower than production which already takes a while with 5 workers.

Reindexing from a remote cluster would probably be faster because our staging and production Elasticsearch clusters are provisioned with the same capacity.

The cut off part of my comment, IIRC, was going to say that the approach to just use the data refresh process for staging is probably faster to implement and could be simplified if we repurposed the existing data refresh DAG to run against multiple environments.

Reply:
Good point! I saw that option of reindex from a remote cluster but I think generally, I'm wary of this kind of data mixing between environments. You make it sound it simple but I cannot see how that would _not_ cause additional load to production. It still has to transfer _all_ the documents from one cluster to another. It may not impede its operation but we certainly do not know.

I also like that the `<media>-full` index product is the result of the same process used for production but just with staging data. If something goes wrong on the way there would be less uncertainty as to what could have gone wrong. I can add this remote cluster approach to the alternatives, maybe we could try as well to evaluate the performance. If it several times faster than using the Ingestion server without degrading the production cluster that would be a big plus to consider leaving it.

Reply:
> You make it sound it simple but I cannot see how that would not cause additional load to production. It still has to transfer all the documents from one cluster to another. It may not impede its operation but we certainly do not know.

To add some additional nuance (but not necessarily arguing strongly in favour of the remote reindex approach, just adding more clarity). The only load I think it could possible cause to production is a full scan through the index documents, which is trivial compared to indexing and even querying, I would think. We can confirm this during a period when production isn't indexing and use something like `max_docs` set to 1 million just to see what happens in production. Reindex jobs are cancellable so it if resource consumption is too high in production for some reason, we can recover very quickly. That's to say: I suspect it will not cause significant load and we can pretty safely find out exactly what that would be. If the main reason to table this alternative (which has the benefits shared before) is because of unknown effects in production, we can confirm whether that's a good reason to table it for sure.

Additionally, now that the staging database is just a restored snapshot of production, there shouldn't be a difference between using the ingestion server worker approach vs reindexing from a remote cluster. The set of documents would be the same and we wouldn't be bottlenecked by the data volume/staging ingestion worker mis-match. That's to say, staging data is just production data. If the intention of this DAG is really to exercise the ingestion server worker approach, though, then that's a different story. That, however, would probably be better suited by retrofitting the existing data refresh DAG to be able to be manually run against staging (and therefore exercising the same data refresh routine that would be used in production) rather than a new DAG that operates slightly differently.

Anyway, I strongly agree with this, of course:

> I can add this remote cluster approach to the alternatives, maybe we could try as well to evaluate the performance. If it several times faster than using the Ingestion server without degrading the production cluster that would be a big plus to consider leaving it.

I'd be very happy to buddy with you on testing this if you'd like.

Reply:
> We can confirm this during a period when production isn't indexing and use something like max_docs set to 1 million just to see what happens in production. Reindex jobs are cancellable so it if resource consumption is too high in production for some reason, we can recover very quickly.

That will be quite an appropriate experiment! Let's give it a shot 🚀

> If the intention of this DAG is really to exercise the ingestion server worker approach, though, then that's a different story. That, however, would probably be better suited by retrofitting the existing data refresh DAG to be able to be manually run against staging (and therefore exercising the same data refresh routine that would be used in production) rather than a new DAG that operates slightly differently.

The thing is that we don't want to run the whole data refresh process, just the part of ES reindexing. That is what the DAG defined here aims for.

Reply:
> That will be quite an appropriate experiment! Let's give it a shot 🚀

Pinged in Slack to coordinate :slightly_smiling_face:

Reply:
**Update**: We tried reindexing from staging to a local ES instance but it wasn't possible to access the staging ES cluster. It required access via HTTPS. The error received was the following:
```json
{
	...
	"error": {
        "type": "illegal_argument_exception",
        "reason": "[localhost:9200] not whitelisted in reindex.remote.whitelist"
    }
}
```

This would require modifications to the cluster settings so we will not pursue this option this time.